### PR TITLE
bpo-39498 Start linking the security warnings in the stdlib modules

### PR DESCRIPTION
**Pull request by tonybaloney**
Within the documentation, there are some really important security considerations for standard library modules. e.g. subprocess, ssl, pickle, xml.

There is currently no "index" of these, so you have to go hunting for them. They're easter eggs within the docs. There isn't a unique admonition type either, so you have to search across many criteria.

In particular for security researchers, it would be useful to consolidate and signpost these security best-practices in one index.

This PR links to some of the existing ones that I found.

https://bugs.python.org/issue39498

### CODE REVIEW

**Review by brandtbucher — APPROVED** [review]
@willingc, what do you think about this?

This looks like a good idea, but there are more modules that I would think to add if we're going this route: `marshal`, `multiprocessing`, and `random`, to name a few.

The PR looks fine to me, but I'm not familiar enough with the security stuff to review the actual content. It looks like we don't have any devs that have registered themselves as "security" experts, but maybe @tiran can weigh in?

**Comment by JulienPalard** [comment]
I also found:
- xmlrpc.server and client warning about not being secure against maliciously constructed data.
- hmac deprecated md5 by default
- hashlib warning about BLAKE2 and some other being unsuitable for hashing passwords
- hashlib warning about hash collision weakness
- logging.config note about portions of the configuration being passed through eval
- pyexpat warning about not begin secure against maliciously constructed data
- shelve warning about using pickle, so being insecure.
- sqlite3 about SQL injections
- tempfile about mktemp not being secure and being deprecated.
- xml, xml.dom.{minidom,pulldom}, xml.etree.ElementTree, xml.sax warning about not being secure against maliciously constructed data

Maybe @vstinner know more, and we may found some other in the https://github.com/pycqa/bandit implementation.

**Comment by vstinner** [comment]
By the way, I'm maintaining https://python-security.readthedocs.io/ website. But I would prefer to migrate it under a .python.org domain before starting to mention it in the official documentation.

**Comment by tonybaloney** [comment]
So between the additions in https://github.com/python/cpython/pull/18272#pullrequestreview-350975899, https://github.com/python/cpython/pull/18272#issuecomment-583148779 and inside the link @vstinner sent (there's a section on dangerous modules and usage)
I can put something together.

I looked at "tagging" them in Sphinx, but it would add a whole bunch of code and engineering complexity that seemed unnecessary.

To keep the scope narrow, this is a "security guidance for standard library modules" index?

**Comment by JulienPalard** [comment]
> To keep the scope narrow, this is a "security guidance for standard library modules" index?

This is a discussion for b.p.o, let's try to avoid discussing on PRs (other than reviewing the commits).

File: Doc/library/security_warnings.rst [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+Security Considerations
+=======================
+
+The following modules have specific security considerations:
+
+* :mod:`ssl` : :ref:`SSL/TLS security considerations <ssl-security>`
+* :mod:`cgi` : :ref:`CGI security considerations <cgi-security>`
+* :mod:`pickle` : :ref:`Restricting globals in pickle <pickle-restrict>`
+* :mod:`subprocess` : :ref:`Subprocess security considerations <subprocess-security>`
+* :mod:`xml` : :ref:`XML vulnerabilities <xml-vulnerabilities>`
```
Comment (vstinner) [inline Doc/library/security_warnings.rst]: There are also:
* https://docs.python.org/dev/library/http.server.html
* logging config uses eval
* hashlib: usedforsecurity=True
* multiprocessing recv uses pickle
* random: "The pseudo-random generators of this module should not be used for security purposes."
* zipfile ZIP bomb: https://docs.python.org/dev/library/zipfile.html#resources-limitations
* os.popen() is vulnerable to shell injection

Maybe this documentation page could be more ambitious and give more hints about security? Examples:

* Discourage the usage of eval(),
* Warn about bugs in Python itself (ex: ast.literal_eval doesn't protect against DoS and bugs in Python parser).
* Windows DLL injection
* sys.path: PYTHONPATH, -I, -E, -s and -S command line options
* Security of embedded Python
* PyPI Supplain Chain

You may even mention tools dedicated to search for Python vulnerabilities, like: https://pypi.org/project/bandit/

See also my notes:

* https://python-security.readthedocs.io/ssl.html
* https://python-security.readthedocs.io/security.html
* https://python-security.readthedocs.io/packages.html
* https://python-security.readthedocs.io/vulnerabilities.html

**Pull request merged**